tests/numerics/smoothness_estimator_02.cc

```diff
@@ -0,0 +1,307 @@
+          << std::endl
+          << linf << std::endl;
+}
+
+template <typename CoefficientType>
+void
+compare(const Table<3, CoefficientType> &coeff1,
+        const Table<3, CoefficientType> &coeff2)
+{
+  for (unsigned int d = 0; d < 3; ++d)
+    Assert(coeff1.size(d) == coeff2.size(d), ExcInternalError());
+
+  typename CoefficientType::value_type linf = 0.;
+  for (unsigned int i = 0; i < coeff1.size(0); i++)
+    for (unsigned int j = 0; j < coeff1.size(1); j++)
+      for (unsigned int k = 0; k < coeff1.size(2); k++)
+        linf = std::max(linf, std::abs(coeff1(i, j, k) - coeff2(i, j, k)));
+
+  deallog << "Linf norm in exact and calculate Fourier coefficients:"
+          << std::endl
+          << linf << std::endl;
+}
+
+
+
+template <int dim>
+void
+test(const unsigned int poly_degree)
+{
+  const unsigned int N        = 4;
+  const unsigned int max_poly = N;
+  deallog << "-----------------------------------" << std::endl;
+  deallog << dim << "d, p=" << poly_degree << ", max_p=" << max_poly
+          << ", N=" << N << std::endl;
+  deallog << "-----------------------------------" << std::endl;
+
+  // --- prepare test function ---
+  PolyFunction<dim> test_function(poly_degree);
+
+  // exact coefficients in 1D case
+  const double &pi  = numbers::PI;
+  const double  pi2 = std::pow(pi, 2);
+  const double  pi3 = std::pow(pi, 3);
+
+  std::vector<std::complex<double>> exact(N);
+  switch (poly_degree)
+    {
+      case 1:
+        exact[0] = std::complex<double>(1., 0.) / 2.;
+        exact[1] = std::complex<double>(0., -1.) / (2. * pi);
+        exact[2] = std::complex<double>(0., -1.) / (4. * pi);
+        exact[3] = std::complex<double>(0., -1.) / (6. * pi);
+        break;
+      case 2:
+        exact[0] = std::complex<double>(1., 0.) / 3.;
+        exact[1] = std::complex<double>(1., -pi) / (2. * pi2);
+        exact[2] = std::complex<double>(1., -2. * pi) / (8. * pi2);
+        exact[3] = std::complex<double>(1., -3. * pi) / (18. * pi2);
+        break;
+      case 3:
+        exact[0] = std::complex<double>(1., 0.) / 4.;
+        exact[1] = std::complex<double>(3. * pi, 3. - 2. * pi2) / (4. * pi3);
+        exact[2] = std::complex<double>(6. * pi, 3. - 8. * pi2) / (32. * pi3);
+        exact[3] = std::complex<double>(3. * pi, 1. - 6. * pi2) / (36. * pi3);
+        break;
+      default:
+        Assert(false, ExcNotImplemented());
+        break;
+    }
+
+  // coefficient table for multi-dimensional case
+  TableIndices<dim> size;
+  for (unsigned int d = 0; d < dim; ++d)
+    size[d] = N;
+  Table<dim, std::complex<double>> coeff_in;
+  coeff_in.reinit(size);
+  prepare_symmetric_coefficients(coeff_in, exact);
+
+
+  // --- prepare data structures ---
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria, 0.0, 1.0); // reference cell
+
+  hp::FECollection<dim> fe_collection;
+  hp::QCollection<dim>  q_collection;
+  QGauss<1>             base_quadrature(2);
```

Comment:
Increasing the number of quadrature points decreases the Linf norm of coefficients successively. I assume that we can consider the calculation of Fourier coefficients as valid. However, we may increase the number of quadrature points in general from `2` to something higher (see also [step-27](https://www.dealii.org/9.0.0/doxygen/deal.II/step_27.html#LaplaceProblemLaplaceProblem)).

However, this does not fix the situation with the Sobolev indices. It appears that the current implementation depends on:
- the number of quadrature points of `base_quadrature` (here: 2)
- the number of copies in the `QIterated` object `quadrature` (here: N)

I'm questioning if my example with the polynomial test functions is suitable at all for this case. Any ideas?

Reply:
The convergence of the Linf norm with the number of quadrature points could also be related to the periodic continuation of the (local) solution, in case steep gradients occur on connecting interfaces, which is the case for this scenario. Correct me when I'm wrong.

Let me check how prone the resulting smoothness indicators are to the choice of quadrature rules. If the indicators are somewhat similar, there is no reason for me to consider this a problem any longer.

However, I would like to know your thoughts on whether you consider an adjustment of the default quadrature rules as important.

Reply:
@davydden  I just noticed that you are using `max_degree+1` as `size_in_each_direction` for the `FESeries::Legendre` objects by default:
https://github.com/dealii/dealii/blob/aae3ae8aface55fb1578857cfaa57d802957fe66/source/numerics/smoothness_estimator.cc#L272
On the other hand, we are just using `max_degree` for `FESeries::Fourier` in `step-27` and here:
https://github.com/dealii/dealii/blob/aae3ae8aface55fb1578857cfaa57d802957fe66/source/numerics/smoothness_estimator.cc#L464

---

Is there a reason behind this? Maybe we should the same parameter for both expansions?

Reply:
- The `+ 1` stems from the fact that both expansion start from the zero-mode. We need to supply this increment for both expansion objects.
- Since `deal.II 8.5.0`, we pick integer multiples of `2pi` instead of `pi` for the `k`-modes of the Fourier series. Thus, we need to double the number of points of the base quadrature from `2` to `4` as well to successfully integrate over a whole period.
- We only use modes up to `max_degree + 1`. This corresponds to the maximal polynomial degree of all used Legendre polynomials. We don't see any benefit from increasing it for the Fourier case.